File: src/main/java/org/codeberg/zenxarch/zombies/entity/effect/StatusEffectZombieEffect.java
package org.codeberg.zenxarch.zombies.entity.effect;

import com.mojang.serialization.MapCodec;
import java.util.List;
import net.minecraft.entity.LivingEntity;
import net.minecraft.entity.effect.StatusEffectInstance;
import net.minecraft.server.world.ServerWorld;
import net.minecraft.util.Util;
import net.minecraft.util.dynamic.Codecs;
import org.codeberg.zenxarch.zombies.entity.ExtendedZombieEntity;
import org.jetbrains.annotations.Nullable;

public record StatusEffectZombieEffect(List<StatusEffectInstance> effects) implements ZombieEffect {

  public static final MapCodec<StatusEffectZombieEffect> CODEC =
      Codecs.listOrSingle(StatusEffectInstance.CODEC)
          .xmap(StatusEffectZombieEffect::new, StatusEffectZombieEffect::effects)
          .fieldOf("effects");

  public StatusEffectZombieEffect(StatusEffectInstance effect) {
    this(List.of(effect));
  }

  @Override
  public void run(
      ServerWorld world, ExtendedZombieEntity zombie, @Nullable LivingEntity adversery) {
    if (adversery == null) return;
    var random = zombie.getRandom();
    var optionalEntry = Util.getRandomOrEmpty(effects, random);
    if (optionalEntry.isPresent()) adversery.addStatusEffect(optionalEntry.get(), zombie);
  }

  @Override
  public MapCodec<StatusEffectZombieEffect> getCodec() {
    return CODEC;
  }
}

File: src/main/java/org/codeberg/zenxarch/zombies/entity/effect/StatusEffectZombieEffect.kt
package org.codeberg.zenxarch.zombies.entity.effect

import net.minecraft.entity.LivingEntity
import net.minecraft.server.world.ServerWorld
import net.minecraft.util.Util
import org.codeberg.zenxarch.zombies.entity.ExtendedZombieEntity
import org.codeberg.zenxarch.zombies.entity.effect.util.StatusEffectInstanceBuilder

@JvmRecord
data class StatusEffectZombieEffect(val effects: List<StatusEffectInstanceBuilder>) : ZombieEffect {
    constructor(effect: StatusEffectInstanceBuilder) : this(listOf<StatusEffectInstanceBuilder>(effect))

    override fun run(
        world: ServerWorld, zombie: ExtendedZombieEntity, adversery: LivingEntity?
    ) {
        if (adversery == null) return
        val random = zombie.random
        val optionalEntry = Util.getRandomOrEmpty(effects, random)
        if (optionalEntry.isPresent) adversery.addStatusEffect(optionalEntry.get().build(), zombie)
    }
}
